[yaml2obj][MachO] Fix crash from integer underflow with invalid cmdsize

rjmansfield opened the pull request:
yaml2obj would crash when processing Mach-O load commands with cmdsize smaller than the actual structure size e.g. LC_SEGMENT_64 with cmdsize=56 instead of 72. The crash occurred due to integer underflow when calculating padding: cmdsize - BytesWritten wraps to a large value when negative, causing a massive allocation attempt.

rjmansfield committed:
[yaml2obj][MachO] Fix crash from integer underflow with invalid cmdsize

yaml2obj would crash when processing Mach-O load commands with cmdsize
smaller than the actual structure size e.g. LC_SEGMENT_64 with
cmdsize=56 instead of 72. The crash occurred due to integer underflow
when calculating padding: cmdsize - BytesWritten wraps to a large value
when negative, causing a massive allocation attempt.

jh7370 commented on llvm/lib/ObjectYAML/MachOEmitter.cpp:
```diff
@@ -285,7 +285,15 @@ void MachOWriter::writeLoadCommands(raw_ostream &OS) {
-    auto BytesRemaining = LC.Data.load_command_data.cmdsize - BytesWritten;
+    // Prevent integer underflow if BytesWritten exceeds cmdsize.
+    if (BytesWritten > LC.Data.load_command_data.cmdsize) {
+      errs() << "warning: load command " << LC.Data.load_command_data.cmd
```
Prefer `WithColor::warning`, same as e.g. https://github.com/rjmansfield/llvm-project/blob/1857805719a0c32263f4356c3902c63d32db15c8/llvm/lib/ObjectYAML/ELFEmitter.cpp#L1555

rjmansfield replied:
Ok, sounds good to me. This is more consistent with llvm-readobj's warning too.

rjmansfield committed:
Address reviewer feedback.

 - Add getLoadCommandName() helper to map cmd values to names
 - Change to indexed loop to show load command index for disambiguation
 - Update warning format to match llvm-readobj style:
    Known commands: "load command 0 LC_SEGMENT_64 cmdsize too small"
    Unknown commands: "load command 0 (0xdeadbeef) cmdsize too small"
 - Use WithColor::warning()

rjmansfield committed:
Address reviewer feedback.
 - Refactor warning message
 - Add testcase for unknown load command

rjmansfield committed:
Fix formatting.

jh7370 commented on llvm/test/ObjectYAML/MachO/load-cmdsize-too-small.yaml:
```diff
@@ -31,3 +36,20 @@ LoadCommands:
     flags:           0
 ...
+
+## Test with an unknown load command value
```
```suggestion
## Test with an unknown load command value.
```

rjmansfield committed:
Address reviewer feedback.
 - Do .str() around the full string concatenation
 - Reorganize test

jh7370 reviewed (approved):
LGTM, assuming pre-merge checks pass.

rjmansfield committed:
Add empty lines before YAML blocks.

rjmansfield commented:
If this change is OK, can someone please merge it on my behalf? Thanks.

jh7370 commented:
> If this change is OK, can someone please merge it on my behalf? Thanks.

Done, sorry for the delay - I've had some time off and not had a chance to go through my review backlog.
